js/background.js

```diff
@@ -98,18 +100,145 @@ chrome.browserAction.onClicked.addListener((tab) =>
+		thumbnails = values?.thumbnails || {};
+		collectionStorage.get(null, values =>
+		{
+			collections = decompressCollectionsStorage(values);
+			UpdateTheme();
+			callback({"collections":collections,"thumbnails":thumbnails})
+		});
+	});
+}
+
+/**
+ * Use a decompression mechanism to decompress stringified collections in an object, one by one.
+ * Ignores non collections items (items with a key not starting with "set_"
+ * @param {Object} compressedCollections object of stringified collections to decompress
+ * @returns {Object} Object of decompressed and parsed collections.
+ */
+function decompressCollectionsStorage(compressedCollections){
+	var decompressedStorage={}
+	for (const [key, value] of Object.entries(compressedCollections)) {
+		if (!key.startsWith("set_"))
+			continue;
+
+		decompressedStorage[key]=JSON.parse(LZUTF8.decompress(value,{inputEncoding:"StorageBinaryString"}));
+		decompressedStorage[key].timestamp=parseInt(key.substr(4))
 	}
-	UpdateTheme();
+	return decompressedStorage
+}
+
+/**
+ * Merges a provided collections array with older pre v2 collections,
+ * saving the result into the new post v2 storage, or into bookmarks on failure.
+ * Allows to preserve backward compatibility with the localStorage method of storing collections in pre v2 versions.
+ * @param {Object} collections The current collections object
+ */
+function MergePreV2Collections({collections}){
+	if (localStorage.getItem("sets"))
+		{
+			console.log("Found pre-v2 data");
+			old_collections=JSON.parse(localStorage.getItem("sets"))
+			//Migrate thumbnails and icons to follow the new format .
+			old_collections.forEach(collection => {
+				for (var i = 0; i < collection.links.length; i++)
+					thumbnails[collection.links[i]]={
+						"pageCapture":collection.thumbnails[i],
+						"iconUrl":collection.icons[i]
+					}
+				delete collection.thumbnails;
+				delete collection.icons;
+
+				UpdateStorages({["set_"+collection.timestamp]:collection},
+					()=> null,
+					()=> {
+						SaveCollectionAsBookmarks(collection);
+						alert(chrome.i18n.getMessage("olderDataMigrationFailed"));
+					});
+			});
+			localStorage.removeItem("sets");
+		}
+}
+
+function SaveCollectionAsBookmarks(collection)
+{
+	//The id 1 is the browser's bookmark bar
+	chrome.bookmarks.create({'parentId': "1",
+			'title': 'TabsAside ' + (collection.name ?? new Date(collection.timestamp).toISOString())
+		}, (collectionFolder) => {
+			for (var i = 0; i < collection.links.length; i++)
+				chrome.bookmarks.create(
+					{'parentId': collectionFolder.id,
+						'title': collection.titles[i],
+						'url': collection.links[i]
+					});
+		});
+}
+
+LoadStorages(MergePreV2Collections)
+
+chrome.storage.onChanged.addListener((changes, namespace) =>
+{
+	if (namespace == "sync")
+		for (key in changes)
+			if (key.startsWith("set_")){
+				if(changes[key].newValue)
+				{
+					collections[key] = decompressCollectionsStorage({[key]: changes[key].newValue})[key];
```

Comment:
Why do we need to wrap `key` into an array?
```javascript
{[key]: changes[key].newValue}
```

Reply:
Oh, it's actually not an array - it's an operator to indicate that the `key` variable will be used as a key, instead of the string `"key"`. 
Some code in the DevTools console should explain the differences better than a lengthy text :

```
>key="AUniqueKey"

>{key:"SomeData"}
   result : {key: "SomeData"}

>{[key]:"SomeData"}
    result : {AUniqueKey: "SomeData"}
```